File: mqtt/golang/consumers/mqtt_consumer.go
package consumers

import (
	"fmt"

	mqtt "github.com/eclipse/paho.mqtt.golang"
)

type MqttConsumer interface {
}

type mqttConsumer struct {
}

func NewMqttConsumer(client mqtt.Client) MqttConsumer {
	topic := "test/topic"
	token := client.Subscribe(topic, 1, func(client mqtt.Client, msg mqtt.Message) {
		fmt.Printf("Pesan diterima: [%s] %s\n", msg.Topic(), msg.Payload())
	})
	token.Wait()
	fmt.Println("subscribed to test/topic")
	return &mqttConsumer{}
}

File: mqtt/rust/src/consumers/mqtt_consumer.rs
use std::{process, sync::Arc};

use chrono::Local;
use rumqttc::{Event, Packet};

use crate::utils::mqtt_util::MqttUtilImpl;

pub struct MqttConsumerImpl {
}

impl MqttConsumerImpl {
    pub async fn new(mqtt_util: Arc<MqttUtilImpl>) -> MqttConsumerImpl {
        println!("{} kafka: subscribing to test/topic", Local::now().format("%Y-%m-%d %H:%M:%S%.3f").to_string());
        match mqtt_util.client.lock().await.subscribe("test/topic", rumqttc::QoS::AtMostOnce).await {
            Ok(_) => {
                println!("{} kafka: subscribed to test/topic", Local::now().format("%Y-%m-%d %H:%M:%S%.3f").to_string());
            },
            Err(err) => {
                println!("{} kafka: cannot subscribe to test/topic", Local::now().format("%Y-%m-%d %H:%M:%S%.3f").to_string());
                println!("error mqtt util: {}", err);
                process::exit(1);
            }
        }

        tokio::spawn(async move {
            while let Ok(event) = mqtt_util.event_loop.lock().await.poll().await {
                if let Event::Incoming(Packet::Publish(p)) = event {
                    println!("{} Received on {}: {}", Local::now().format("%Y-%m-%d %H:%M:%S%.3f").to_string(), p.topic, String::from_utf8_lossy(&p.payload));
                }
            }
        });

        MqttConsumerImpl { 
        }
    }
}